url/url.h

```diff
@@ -31,10 +38,227 @@ struct Origin {
 };
 
 /**
- * Generates a new Blob URL for the given origin
+ * Generates a new Blob URL for the given origin.
  */
 std::string blob_url_create(Origin const &origin);
 
+struct Url {
+    std::string scheme;
+    std::string user;
+    std::string passwd;
+    std::optional<Host> host;
+    std::optional<std::uint16_t> port;
+    std::variant<std::string, std::vector<std::string>> path;
+    std::optional<std::string> query;
+    std::optional<std::string> fragment;
+};
+
+// This parser is current with the WHATWG URL specification as of 1 March 2023
+class UrlParser final : util::BaseParser {
+public:
+    UrlParser() : BaseParser{""} {}
+
+    std::optional<Url> parse(
+            std::string input, std::optional<Url> base = std::nullopt, std::optional<std::string> encoding = "UTF8");
+
+private:
+    enum class ParserState {
+        SchemeStart,
+        Scheme,
+        NoScheme,
+        SpecialRelativeOrAuthority,
+        PathOrAuthority,
+        Relative,
+        RelativeSlash,
+        SpecialAuthoritySlashes,
+        SpecialAuthorityIgnoreSlashes,
+        Authority,
+        Host,
+        Hostname,
+        Port,
+        File,
+        FileSlash,
+        FileHost,
+        PathStart,
+        Path,
+        OpaquePath,
+        Query,
+        Fragment,
+        Failure,
+        Terminate
+    };
+
+    enum class ValidationError {
+        // IDNA
+        DomainToASCII,
+        DomainToUnicode,
+        // Host parsing
+        DomainInvalidCodePoint,
+        HostInvalidCodePoint,
+        IPv4EmptyPart,
+        IPv4TooManyParts,
+        IPv4NonNumericPart,
+        IPv4NonDecimalPart,
+        IPv4OutOfRangePart,
+        IPv6Unclosed,
+        IPv6InvalidCompression,
+        IPv6TooManyPieces,
+        IPv6MultipleCompression,
+        IPv6InvalidCodePoint,
+        IPv6TooFewPieces,
+        IPv4InIPv6TooManyPieces,
+        IPv4InIPv6InvalidCodePoint,
+        IPv4InIPv6OutOfRangePart,
+        IPv4InIPv6TooFewParts,
+        // URL parsing
+        InvalidURLUnit,
+        SpecialSchemeMissingFollowingSolidus,
+        MissingSchemeNonRelativeURL,
+        InvalidReverseSolidus,
+        InvalidCredentials,
+        HostMissing,
+        PortOutOfRange,
+        PortInvalid,
+        FileInvalidWindowsDriveLetter,
+        FileInvalidWindowsDriveLetterHost
+    };
+
+    const std::map<ValidationError, std::string> validation_error_str = {
```

Comment:
Looks like this, the special_schemes, and the PercentEncodeSet can go in the .cpp since they're only used internally?

Reply:
Every private member and method of this class is only used internally. I'm not sure I understand why they should be moved. If I do that, what is the idiomatic way to move these out of the header? Like, do I remove them from the class entirely?

Reply:
The private members are fine, but things like this massive enum -> string map are mostly noise in the header and doesn't help you get an overview of what the class does. Usually I'd just dump that kind of thing at the top of the .cpp in an anonymous namespace, or close to where it's used.

Maybe the PercentEncodeSet should go together with the function in util? Would make testing it a lot easier as well.